src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/AsyncCollectorBindingProvider.cs

```diff
@@ -0,0 +1,303 @@
+                {
+                    return new BindingMode(Mode.ICollector, elementType);
+                }
+
+                // A different interface. Let another rule try it. 
+                return null;
+            }
+
+            if (parameter.IsOut)
+            {
+                // How should "out byte[]" bind?
+                // If there's an explicit "byte[] --> TMessage" converter, then that takes precedence.                 
+                // Else, bind over an array of "byte --> TMessage" converters 
+                Type elementType = parameter.ParameterType.GetElementType();
+                bool hasConverter = TypeUtility.HasConverter<TAttribute>(this._converterManager, elementType, typeof(TType));
+                if (hasConverter)
+                {
+                    // out T, where T might be an array 
+                    return new BindingMode(Mode.OutSingle, elementType);
+                }
+
+                if (elementType.IsArray)
+                {
+                    // out T[]
+                    var messageType = elementType.GetElementType();
+                    return new BindingMode(Mode.OutArray, messageType);
+                }
+
+                var checker = ConverterManager.GetTypeValidator<TType>();
+                if (checker.IsMatch(elementType))
+                {
+                    // out T, t is not an array 
+                    return new BindingMode(Mode.OutSingle, elementType);                    
+                }
+
+                // For out-param ,we don't expect another rule to claim it. So give some rich errors on mismatch.
+                if (typeof(IEnumerable).IsAssignableFrom(elementType))
+                {
+                    throw new InvalidOperationException(
+                        "Enumerable types are not supported. Use ICollector<T> or IAsyncCollector<T> instead.");
+                }
+                else if (typeof(object) == elementType)
+                {
+                    throw new InvalidOperationException("Object element types are not supported.");
+                }
+            }
+
+            // No match. Let another rule claim it
+            return null;            
+        }
+
+        // Represent the different possible flavors for binding to an async collector
+        private class BindingMode
+        {
+            public BindingMode(Mode mode, Type elementType)
+            {
+                this.Mode = mode;
+                this.ElementType = elementType;
+            }
+            public Mode Mode { get; set; }
+            public Type ElementType { get; set; }
+        }
+
+        // TType - specified in the rule. 
+        // TMessage - extracted from the user's parameter. 
+        private class ExactBinding<TMessage> : BindingBase<TAttribute>
+        {
+            private readonly Func<object, object> _buildFromAttribute;
+
+            private readonly FuncConverter<TMessage, TAttribute, TType> _converter;
```

Comment:
I think this rename is out of scope. It's a pre-existing type that's orthogonal to this change.